Use native rust-bitcoin methods for weight calculations

spacebear21 opened the pull request:
See https://github.com/payjoin/rust-payjoin/issues/353

This deletes the custom `weight.rs` traits and `input_type.rs` helpers in favor of out-of-the-box rust-bitcoin methods. It also removes the unused `output_type.rs`.

A question that may be relevant to this PR: do we want to commit to allowing mixed input types (despite it being a violation of the BIP78 spec) to better enable batching use cases? Maybe we could introduce a `allow_mixed_input_types` parameter that can be set by the sender and defaults to false if not set?

DanGould commented:
pretty massive delete! eager to follow

DanGould commented:
I'm partial to allow mixed input types, and the [consensus seems to be agreement](https://github.com/bitcoin/bips/pull/1244). In addition, allow by default is more permissive and should have fewer failed attempts, I'm not sure I'd even want to expose the option unless it were strictly necessary for a downstream usecase *if the logic weren't already in the codebase* for v1.

spacebear21 commented on payjoin/src/input_type.rs:
```diff
@@ -128,18 +132,30 @@ impl InputType {
+            // witness: <signature> <pubkey> = 72, 33 bytes
+            // https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#p2wpkh-nested-in-bip16-p2sh
+                InputWeightPrediction::new(23, &[72, 33]),
             SegWitV0 { ty: SegWitV0Type::Script, nested: _ } => unimplemented!(),
```
This clause applies to P2SH-P2WSH. The suggestion for P2SH-P2WPKH is implemented just above it with `InputWeightPrediction::new(23, &[72, 33])` which corresponds to the suggested 91 vsize after adding the txid, index, and sequence bytes.

spacebear21 commented on payjoin/src/psbt.rs:
```diff
@@ -180,6 +190,44 @@ impl<'a> InputPair<'a> {
+                    // input script: 0x160014{20-byte-key-hash} = 23 bytes
+                    // witness: <signature> <pubkey> = 72, 33 bytes
+                    // https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#p2wpkh-nested-in-bip16-p2sh
+                        InputWeightPrediction::new(23, &[72, 33]),
```
Opened a PR in rust-bitcoin to upstream it: https://github.com/rust-bitcoin/rust-bitcoin/pull/3443

spacebear21 commented on payjoin/src/send/mod.rs:
```diff
@@ -125,26 +127,28 @@ impl<'a> RequestBuilder<'a> {
+                .iter()
+                .try_fold(true, |_, input_pair| -> Result<bool, crate::psbt::AddressTypeError> {
+                    Ok(input_pair.address_type()? == first_input_pair.address_type()?)
                 })
```
Good catch. I rewrote this as a simple `while` loop which breaks if an input is not the same type as the first one. That seems easier to reason about.

DanGould replied:
The new while loop is way easier to reason about

I can't recall where the rule to // use cheapest default if mixed input types` (P2TR) estimation method comes from when mixed input types are present. Wouldn't it make more sense to use the cheapest *of the scripts proposed*? As long as we make it easy for another compatible implementation to come along, I don't think it so much matters either way.

spacebear21 commented:
@DanGould Summary of changes since your last review:
- Fixed the `try_fold` issue you brought up.
- Removed the `input_type` and `sequence` fields from the ContextV1 and RequestContext structs, as they can be computed from the original psbt.
- Added a unit test for input weight calculations. I used payjoin-cli with `RUST_LOG=trace` to generate and obtain the original and payjoin PSBT values, and a new bitcoin-cli wallet for each address type.

DanGould commented:
- Try_fold fix looks good
- input_type and sequence removal also looking good

I haven't checked the PSBT values in the known-weight tests, so I'm on that tomorrow. I think that's the last substantive check to make.

DanGould reviewed (approved):
I verified the additional input weight comes from scripts of the stated address types. The UTXO locking scripts are as follows and seem to be the the specified AddressType
- `p2pkh_proposal` adds `vout: 0` having p2pkh script `OP_DUP OP_HASH160 OP_PUSHBYTES_20 3d786edc8e83f51d30529bd3bef526f9518d70db OP_EQUALVERIFY OP_CHECKSIG`
- `nested_p2wpkh_proposal` adds `vout: 0` having p2wpkh script `OP_HASH160 OP_PUSHBYTES_20 c7ff991c104167ee84fd44b537639eec80c8b578 OP_EQUAL`
- `p2wpkh_proposal` adds `vout: 0` having p2wpkh script `OP_0 OP_PUSHBYTES_20 916d91f1d3b77cb510331a37a0858e3356b69091`
- `p2tr_proposal` adds `vout: 0` having p2tr script `OP_PUSHNUM_1 OP_PUSHBYTES_32 9f6db5fe14725bcc2ef578414bb0c8c3a2e3096eef2881e47478da0dc4c33d0b`

The only nits I'd ask for are for the warnings of unused variables to be removed from 

- secp256k1 dependencies in send/mod.rs
- `let sequence = original_psbt.unsigned_tx.input[0].sequence;` in send/mod.rs

and for the `input_weight_calculations` test name to be `additional_input_matches_known_weight` or similar boolean statement

DanGould commented on payjoin/src/receive/mod.rs:
```diff
@@ -229,14 +228,8 @@ impl MaybeMixedInputScripts {
-                    }
-                },
+            .scan(&mut err, |err, input| match Ok(input.address_type()) {
+                Ok(address_type) => Some(address_type),
```
Seems like this will never reach the `Err(e)` branch since match is directly called on an `Ok` value?

spacebear21 replied:
Yes - this was a temporary workaround before I had `address_type()` return a `Result`. It gets updated in https://github.com/payjoin/rust-payjoin/pull/360/commits/6938b70f486a3999063fb2218766212ee5855c99#diff-5fba6fc39c59b07e5980d87da92537138068364d192a2b63eee5e57af01a07bb

DanGould commented on payjoin/src/send/mod.rs:
```diff
@@ -561,6 +542,11 @@ impl ContextV1 {
+                        .original_psbt
+                        .input_pairs()
+                        .next()
+                        .expect("original PSBT should have an input");
```
Can we handle this error?

spacebear21 replied:
It would mean the original PSBT had no inputs, which seems improbable at this stage since the receiver would've accepted it and responded with a payjoin proposal. I could add a proper error type for it though.

spacebear21 replied:
Added proper handling for this error in 68fc635f57bd
